lib/stage.ts

```diff
@@ -0,0 +1,231 @@
+	StageDependentActionGroup,
+} from './action-group';
+import { DockerfileParseError } from './errors';
+
+export class Stage {
+	public name: string;
+	public index: number;
+	public dependentOnStages: number[] = [];
+	public isLast: boolean = false;
+	public actionGroups: ActionGroup[] = [
+		{
+			dependentOnStage: false,
+			commands: [],
+			copies: [],
+			workdir: '/',
+		},
+	];
+
+	/*
+		This var is used to detect whether we create a new action group,
+		as we can collect as many copies together into the same action group,
+		but to ensure we run the commands in the correct order, a new copy forces
+		a new group
+	*/
+	private lastStepWasCopy: boolean = false;
+	private lastWorkdir: string = '/';
+	private ungroupedCommands: Command[] = [];
+
+	public constructor(idx: number, name?: string) {
+		this.index = idx;
+		this.name = name != null ? name : idx.toString();
+	}
+
+	public addLocalCopyStep(args: string[]) {
+		const lastActionGroup = _.last(
+			this.actionGroups,
+		) as LocalDependentActionGroup;
+
+		if (this.canAddCopyToGroup(false)) {
+			lastActionGroup.copies = lastActionGroup.copies.concat(
+				this.copyArgsToCopies(args, lastActionGroup),
+			);
+		} else {
+			lastActionGroup.commands = lastActionGroup.commands.concat(
+				this.ungroupedCommands,
+			);
+			const actionGroup: LocalDependentActionGroup = {
+				dependentOnStage: false,
+				commands: [],
+				copies: [],
+				workdir: this.lastWorkdir,
+			};
+			actionGroup.copies = this.copyArgsToCopies(args, actionGroup);
+			this.actionGroups.push(actionGroup);
+			this.ungroupedCommands = [];
+		}
+
+		this.lastStepWasCopy = true;
+	}
+
+	public addStageCopyStep(args: string[], stageIdx: number) {
+		const lastActionGroup = _.last(
+			this.actionGroups,
+		) as StageDependentActionGroup;
+		if (this.canAddCopyToGroup(true, stageIdx)) {
+			lastActionGroup.copies.concat(
+				this.copyArgsToCopies(args, lastActionGroup).map(copy => ({
+					sourceStage: stageIdx,
+					...copy,
+				})),
+			);
+		} else {
+			lastActionGroup.commands = lastActionGroup.commands.concat(
+				this.ungroupedCommands,
+			);
+			const actionGroup: StageDependentActionGroup = {
+				dependentOnStage: true,
+				commands: [],
+				stageDependency: stageIdx,
+				copies: [],
+				workdir: this.lastWorkdir,
+			};
+
+			actionGroup.copies = this.copyArgsToCopies(args, actionGroup).map(
+				copy => ({
+					sourceStage: stageIdx,
+					...copy,
+				}),
+			);
+
+			this.actionGroups.push(actionGroup);
+			this.ungroupedCommands = [];
+		}
+
+		this.lastStepWasCopy = true;
+
+		this.dependentOnStages = _.uniq(this.dependentOnStages.concat(stageIdx));
+	}
+
+	public addCommandStep(command: string) {
+		this.ungroupedCommands.push(command);
+		this.lastStepWasCopy = false;
+	}
+
+	public addWorkdirStep(workdir: string) {
+		// We need to create a new group, saving any ungrouped commands into
+		// the latest group
+		const actionGroup = _.last(this.actionGroups) as ActionGroup;
```

Comment:
I think you can get rid of this cast with
```ts
if (actionGroup == null) {
	throw new Error(...)
}
```
as that should narrow the type correctly and guarantee the type is what we expect

Reply:
I did think about that, but we know for certain that it's always going to exist. I feel like adding this to the code says to the reader that for some reason they cannot be sure it's going to exist.

Reply:
At the very least it should be
```suggestion
		const actionGroup = _.last(this.actionGroups)!;
```
to make it clear all you're doing is excluding nulls